File: Leetcode_Medium/_475. Heaters/Solution.py
class Solution(object):
    def findRadius(self, houses, heaters):
        """
        :type houses: List[int]
        :type heaters: List[int]
        :rtype: int
        """
        # We first sort the two array so 
        # that we can search through it more easily
        houses.sort()
        heaters.sort()
        # The smaller one will be the radius we need
        # Intialize the i index to indicate
        # the current heater position
        # and the current radius
        i, radius = 0, 0
        for house in houses:
            # Move to the headers that's on the right side
            # or at the exact position of the house
            while i < len(heaters) and heaters[i] <= house:
                i+= 1
            # The left indicate the distance to the 
            # nearest left side heater
            left = house - heaters[i-1]
            # The right indicate the distance to the 
            # nearest right side heater
            right = heaters[i] - house

            # Update the radius to become the bigger value
            # between the previous radius and the smaller 
            # distance to reach the current house
            radius = max(radius, min(left,right))
        return radius

        
        

            
        

File: Leetcode_Medium/_475. Heaters/Solution.java
import java.util.Arrays;

public class Solution {
    public int findRadius(int[] houses, int[] heaters) {
        // Sort the houses and heaters arrays
        Arrays.sort(houses);
        Arrays.sort(heaters);
        
        int radius = 0;
        int i = 0; // index for houses
        int j = 0; // index for heaters
        
        // Iterate through each house
        while (i < houses.length) {
            // Find the closest heater to the current house
            while (j < heaters.length - 1 &&
                   Math.abs(heaters[j] - houses[i]) >= Math.abs(heaters[j + 1] - houses[i])) {
                j++;
            }
            // Update the radius to cover the current house
            radius = Math.max(radius, Math.abs(heaters[j] - houses[i]));
            i++;
        }
        
        return radius;
    }
}
